src/main/blackbox/blackbox.c

```diff
@@ -1250,14 +1250,22 @@ static void loadMainState(timeUs_t currentTimeUs)
 
 #if defined(USE_ACC)
         blackboxCurrent->accADC[i] = lrintf(acc.accADC.v[i]);
-        STATIC_ASSERT(offsetof(quaternion_t, w) == 0, "Code expects quaternion in w, x, y, z order");
-        blackboxCurrent->imuAttitudeQuaternion[i] = lrintf(imuAttitudeQuaternion.v[i + 1] * 0x7FFF);  //Scale to int16 by value 0x7FFF = 2^15 - 1; Use i+1 index for x,y,z components access, [0] - w
 #endif
 #ifdef USE_MAG
         blackboxCurrent->magADC[i] = lrintf(mag.magADC.v[i]);
 #endif
     }
-
+#if defined(USE_ACC) // IMU quaternion
+    {
+        // write x,y,z of IMU quaternion. Make sure that w is always positive
+        STATIC_ASSERT(offsetof(quaternion_t, w) == 0, "Code expects quaternion in w, x, y, z order");
+        const float q_sign = imuAttitudeQuaternion.w < 0 ? -1 : 1;     // invert quaternion if w is negative
+        const float q_scale = q_sign * 0x7FFF;                         // Scale to int16 by value 0x7FFF = 2^15 - 1 (-1 <= x,y,z <= 1)
```

Comment:
BF has a sign function we could use (unless the case for w=0 is a problem)

![image](https://github.com/user-attachments/assets/3b702050-b2f5-4dc8-973e-5e78e1075fbb)
```suggestion
        const float q_sign = SIGN(imuAttitudeQuaternion.w);            // invert quaternion if w is negative
        const float q_scale = q_sign * 0x7FFF;                         // Scale to int16 by value 0x7FFF = 2^15 - 1 (-1 <= x,y,z <= 1)
```

Reply:
It won't work here. With w=0 vector (x, y, z) is nonzero, using sign function would reset it to zero (from 180 deg rotation to 0 deg)

Reply:
There is sharp discountuity in reconstructed quaternion when w crosses zero. But both represent identical orientation.